src/common/cockpitjson.c

```diff
@@ -967,3 +967,165 @@ cockpit_json_to_hash_table (JsonObject *object,
 
   return hash_table;
 }
+
+typedef const struct {
+  CockpitJsonWalkCallback callback;
+  gpointer                user_data;
+} WalkClosure;
+
+static JsonNode *
+cockpit_json_walk_node (JsonNode    *node,
+                        WalkClosure *closure);
+
+static JsonArray *
+cockpit_json_walk_array (JsonArray   *array,
+                         WalkClosure *closure)
+{
+  gint n = json_array_get_length (array);
+  JsonArray *copy = NULL;
+
+  for (gint i = 0; i < n; i++)
+    {
+      JsonNode *element = json_array_get_element (array, i);
+      JsonNode *node = cockpit_json_walk_node (element, closure);
+
+      if (node && !copy)
+        {
+          /* We've decided we need to copy now, so revisit everything up
+           * to this point and add it to the copy.
+           */
+          copy = json_array_sized_new (n);
+          for (gint j = 0; j < i; j++)
+            json_array_add_element (copy, json_array_dup_element (array, j));
+        }
+
+      if (copy)
+        json_array_add_element (copy, node ?: json_node_ref (element));
+    }
+
+  return copy;
+}
+
+static JsonObject *
+cockpit_json_walk_object (JsonObject  *object,
+                          WalkClosure *closure)
+{
+  JsonObject *copy = NULL;
+  JsonObjectIter iter;
+  const gchar *key;
+  JsonNode *value;
+
+  json_object_iter_init (&iter, object);
+  while (json_object_iter_next (&iter, &key, &value))
+    {
+      JsonNode *node = cockpit_json_walk_node (value, closure);
+
+      if (node && !copy)
+        {
+          const gchar *this_key = key;
+
+          /* We've decided we need to copy now, so revisit everything up
+           * to this point and add it to the copy.
+           */
+          copy = json_object_new ();
+          json_object_iter_init (&iter, object);
+          while (json_object_iter_next (&iter, &key, &value) && key != this_key)
+            json_object_set_member (copy, key, json_node_ref (value));
+
+          /* ...now key and value will be where they were before we entered */
+          g_assert (key == this_key);
```

Comment:
This is not a tautology.  The loop could also exit because we hit the end of the iter, and that's the case I wanted to guard against.

I actually previously kept a counter here (as with the array case) and iterated for that many counts, but I found it made the code hairy and hard to read.... I'd be happy to discuss this with you further, and maybe we can figure out a way to give you better reassurances about this...

Reply:
Argh, of course, I mis-read the condition. Thanks for considering, let's not make the code unnecessarily complicated then.